Partial executer

carlopi opened the pull request.

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -210,6 +210,9 @@ class PartialInterpreter : public llvm::Interpreter {
 		if (const GlobalVariable* GV = dyn_cast<GlobalVariable>(V))
 		{
+			//Force alignment on queryed globals, this improves the information gained
+			(const_cast<GlobalVariable*>(GV))->setAlignment(Align(8));
```
Unneeded parenthesis

alexp-sssup commented on clang/lib/Driver/ToolChains/WebAssembly.cpp:
```diff
@@ -572,6 +572,7 @@ void cheerp::CheerpOptimizer::ConstructJob(Compilation &C, const JobAction &JA,
     CmdArgs.push_back("-FreeAndDeleteRemoval");
     CmdArgs.push_back("-cheerp-lto");
     CmdArgs.push_back("-Os");
+    CmdArgs.push_back("-PartialExecuter");
```
This commit should be ordered _after_ the pass is actually introduced

carlopi replied:
ok

alexp-sssup commented on llvm/cmake/modules/HandleLLVMOptions.cmake:
```diff
@@ -149,7 +149,7 @@ else(WIN32)
   else(FUCHSIA OR UNIX)
     MESSAGE(SEND_ERROR "Unable to determine platform")
-  endif(FUCHSIA OR UNIX)
+  endif(FUCHSIA OR UNIX OR CHEERP)
```
What's the purpose of this?

carlopi replied:
Fix a warning: CMakes says that connected to this line:
if(FUCHSIA OR UNIX OR CHEERP)
saying that the endif-clause do not match the opening clause.
This is minor, and arguably the same fix should also go into line 150.

alexp-sssup commented on llvm/unittests/Cheerp/CheerpPointerAnalyzerTest.cpp:
```diff
@@ -36,7 +36,7 @@ TEST(CheerpTest, PointerAnalyzerTest) {
 	SMDiagnostic Err;
 
-	std::unique_ptr<Module> M = parseIRFile( "test/test1.ll", Err, C );
+	std::unique_ptr<Module> M = parseIRFile( "test1.ll", Err, C );
```
This is not correct, when the tests are fully run the path will be the existing one.

alexp-sssup replied:
The current path works when executed by jenkins, that is the only relevant case.

carlopi replied:
Sure, dropping this

carlopi replied:
Added -v to the test command: ninja check executes things from a different folder than ninja check-llvm, jenkins uses the former while I manually used the latter. Dropping this commit. The proper fix would be modifying llvm/unittests/Cheerp/CMakeLists somehow to have test.ll in the right place independently of where the command is invoked

alexp-sssup commented on llvm/lib/ExecutionEngine/ExecutionEngine.cpp:
```diff
@@ -1356,6 +1356,12 @@ void ExecutionEngine::emitGlobalVariable(const GlobalVariable *GV) {
   // Don't initialize if it's thread local, let the client do it.
   if (!GV->isThreadLocal()) {
     if (!GV->hasInitializer()) {
+      if (!isForPreExecute) {
```
2 things:

1) There is no point in having 2 identical flags in both the base and derived class. The flag from the derived class should be moved in the base one.
2) I am not sure this whole thing is sound. What is the purpose of this?

carlopi replied:
While PreExecuting, everything has to be intialized, so this commit:
https://github.com/leaningtech/cheerp-compiler/commit/13371bd3b865739391e61e27484935b85b1c88ef
adds the checks for debugging purposes.
Now when NON preexecuting, I think it's fine to just return, or even go to the end of the function.

alexp-sssup commented on llvm/lib/ExecutionEngine/Interpreter/Execution.cpp:
```diff
@@ -336,6 +336,16 @@ void Interpreter::visitICmpInst(ICmpInst &I) {
   GenericValue Src1 = getOperandValue(I.getOperand(0), SF);
   GenericValue Src2 = getOperandValue(I.getOperand(1), SF);
+
+  if (Src1.IntVal.getBitWidth() < Src2.IntVal.getBitWidth())
```
This cannot ever happen. The bitwidth of cmp operands are always identical.

alexp-sssup replied:
I don't understand. IntVal themselves store the bitwidth internally. How could they ever be mismatched? The IR itself would be broken.

carlopi replied:
GenericValue here is calculated by the LLVM's Interpreter, it's possible that it doesn't conform completely with the IR (expecially regarding bit-witdh). I will reproduce and come back

alexp-sssup commented on llvm/lib/CheerpUtils/CMakeLists.txt:
```diff
@@ -28,6 +28,10 @@ add_llvm_component_library(LLVMCheerpUtils
   StoreMerging.cpp
   CheerpLowerInvoke.cpp
   SinkGenerator.cpp
+  PartialExecuter.cpp
```
Merge this commit with the actual one adding the .cpp file, otherwise we have a state which will not build

alexp-sssup replied:
By the way, I am quite sure we had to move PreExecute into CheerpWriter due to weird dependencies issues. Please move PartialExecuter there as well.

alexp-sssup commented on llvm/include/llvm/ExecutionEngine/ExecutionEngine.h:
```diff
@@ -555,6 +555,9 @@ class ExecutionEngine {
   void emitGlobalVariable(const GlobalVariable *GV);
 
+  // Helper function to access protected getConstantValue
+  static GenericValue getConstantValue(ExecutionEngine& ee, const Constant* C);
```
Is this actually ever used?

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1421 @@
+						argIndex++;
+					}
+
+					addStackFrame();
```
That 'hasToBeSkipped" also populate the stack frame does not make much sense. This method should only make decision, not act on them.

carlopi replied:
ok, will split the decision vs populating values

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1421 @@
+
+						stronglyKnownBits.pop_back();
+						if (retVal)
+							stronglyKnownBits.back()[caller] = BITMASK;
```
This does not look correct. caller is the function, not the return values. Why are we setting the bitmask for the function itself?

alexp-sssup replied:
Or is caller the CallInst?

carlopi replied:
Clarified

alexp-sssup commented on llvm/lib/ExecutionEngine/Interpreter/Interpreter.h:
```diff
@@ -93,6 +93,47 @@ class Interpreter : public ExecutionEngine, public InstVisitor<Interpreter> {
+    assert(ECStack.size() > 1);
+    return ECStack[ECStack.size() - 2];
+  }
+  ExecutionContext& getFirstStack()
```
Unused?

carlopi commented on llvm/lib/ExecutionEngine/Interpreter/Interpreter.h:
```diff
@@ -93,6 +93,47 @@ class Interpreter : public ExecutionEngine, public InstVisitor<Interpreter> {
+        return true;
+    return false;
+  }
+  bool isBottomStack() const
```
isInitialStack maybe? this is the execution stack of the function we are currently partially execution (while entry point to me means the actual callsite)
Actually stack is probably the wrong terminology, since ECStack is a stack of execution contexts, so probably all these functions could be named isFrontExecutionContext or isFrontExecContext

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		return false;
+	if (const GlobalVariable* GVar = dyn_cast<GlobalVariable>(V))
+	{
+		if (!GVar->hasInitializer() || GVar->isExternallyInitialized())
```
Either 3 ifs or 1 if with 3 conditions. This is a confusing mix

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	friend FunctionData;
+	friend ModuleData;
+	const llvm::BasicBlock* fromBB{nullptr};	
+	enum BitMask : uint32_t
```
I have doubts about the behavior of 64-bit integers, are those taken into account?

alexp-sssup replied:
I may advise to use a shorter value such as ALL = 0xff, then. The idea would be that the current mask _might_ suggest this to be 32-bit specialized, while it's not.

carlopi replied:
Perfect, done

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	std::vector<const llvm::PHINode*> notComputedPhis;
+	std::vector<std::pair<long long,long long>> validLoadIntervals;
+public:
+	GenericValue getOperandValueExplicit(const Value* V, ExecutionContext& EC)
```
Is this function only used to apply a const-cast? Then it's better to not use constantness on our side

carlopi replied:
Yes, this is to wrap the ExecutionEngine code that don't use const inside it's data structures (but values are const in some sense since we will not call any modifying instruction).
Removing (basically) all const our side also works.

alexp-sssup replied:
I'd say remove constness on our side where the engine does not agree. const_casts suggest that _we_ are messing up with assumptions, which is not the case.

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		for (const llvm::BasicBlock* bb : inner)
+		{
+			data.currentEE->setIncomingBB(nullptr);
+			data.currentEE->visitBasicBlock<PartialInterpreter::VisitingPolicy::REMOVE_VALUES>(bb, nullptr);
```
I don't fully understand the needs for policies here. We need to remove values from maps, independently of their actual execution. Can't we simply iterate on instructions?

carlopi replied:
Right, will remove them altogether and add a loop over instructions (since now they are equivalent). (had some meaning in the old version, since visitBasicBlock was also visiting other blocks, but now it's only ever just removing values and returning)

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+				[[clang::fallthrough]];
+			}
+			case Instruction::Or:
+			case Instruction::Xor:
```
Merge with block above?

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+			}
+		}
+	}
+	void setIncomingBB(const llvm::BasicBlock* from)
```
This setter/getter are only used in 1 place, and also names just like an internal LLVM function. Let's just use the variable.

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	}
+	enum VisitingPolicy {COMPUTE_VALUES, REMOVE_VALUES};
+	template <VisitingPolicy policy>
+	const llvm::BasicBlock* visitBasicBlock(const llvm::BasicBlock* BB, const llvm::BasicBlock* from=nullptr);
```
The default value is never used.

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		: llvm::Interpreter(std::move(M), /*preExecute*/false)
+	{
+	}
+	BitMask getBitMask(const llvm::Value* V) const
```
Should this be merged with computeStronglyKnownBits?

carlopi replied:
Even though they are intertwined, I will review this

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+			return true;
+		if (isa<Argument>(V) || isa<Instruction>(V))
+		{
+			return stronglyKnownBits.count(V);
```
Single line, no { }

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		if (const BranchInst* SI = dyn_cast <BranchInst>(&I))
+		{
+			if (SI->isConditional())
+				if ((getBitMask(SI->getCondition()) != BitMask::ALL) )
```
The condition is i1, so there is no point in checking the bitmask.

carlopi replied:
getBitMask will return NONE in some patological cases (actually == NONE it's clearer)

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+			default:
+			{
+				// TODO(carlo): review PtrToInt or IntToPtr
+				// TODO(carlo): actually the proper way would be have the default meaning it has to be skipped, so that the conservative answer is failure
```
This comment seems important enough to be dealt with.

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+				continue;
+			if (!GV->hasInitializer())
+				continue;
+			if (GV->isExternallyInitialized())
```
Please review this usage for isExternallyInitialized. We check for internal linkage anway. How can an internal variable be externally initialized?

carlopi replied:
I removed the condition (putting an assertion), since I am not actually completely sure of the relations between those + the fact that hasInternalLinkage might get relaxed at some point

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+				continue;
+			if (!GV->hasInternalLinkage())
+				continue;
+			if (!GV->hasName())
```
Why does this matter?

carlopi replied:
I will add a comment, the Interpreter will possibly use internally queries on the name -> failure otherwise

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+			GenericValue SRC = getOperandValue(GV);
+			GenericValue *Ptr = (GenericValue*)GVTORP(SRC);
+			
+			validLoadIntervals.push_back({(long long)Ptr, ((long long)Ptr)+getDataLayout().getTypeAllocSize(GV->getInitializer()->getType())});
```
validLoadIntervals -> constantLoadIntervals? immutableLoadIntervals?

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	template <PartialInterpreter::VisitingPolicy policy>
+	void visitOuter(const llvm::Instruction& I) 
+	{
+		if (policy == REMOVE_VALUES)
```
Quite sure this policy system should be removed.

carlopi replied:
Removed

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		}
+		else
+		{
+			//While visiting the first NonPhi instruction we set the maps state correctly
```
Please make it explicit that this is required to deal with the simultaneity of PHIs. By the way, it seems weird we need to special case this. The exact same problem applies to the normal interpreter (i.e. PHIs needs to be dealt with all together, and not one-by-one). May there be a better entry point for PHIs?

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+
+		if (isBottomStack())
+		{
+			if (I.isTerminator())
```
Hmmmm... Can't this be determined by hasToBeSkipped?

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		else
+		{
+			//Terminators have special handling done here explicitly
+			if (I.isTerminator())
```
Reduce is/else depth by using else if

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+			}
+
+			//We are inside a call, here we assume all failure to execute are non-recoverable (as in no information could be gained)
+			if (skip)
```
It should be possible to move this check above, and only reason over the various terminators if 'skip' is false. I believe that if skip is false then it is guaranteed that the terminators can be resolved and 'next' will be known.

carlopi replied:
Reworked, should be clearer now (factored out the logic for finding the next BB)

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	{
+		//First count how many edges there are betwen from and to
+		uint32_t cardinality = 0;
+		for (BasicBlock* bb : predecessors(to))
```
I think iterating on from successors is cleaner. Also, use a lambda to get the cardinality value instead of this free standing blocks.

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+				else
+				{
+					BI->eraseFromParent();
+					new UnreachableInst(from->getParent()->getParent()->getContext(), from);
```
Context is available everywhere in the hierarchy, not need for the getParent calls.

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+			}
+			case Instruction::Unreachable:
+			{
+				return;
```
This can never happen?

carlopi replied:
It might have been generated by a previous iteration, in degenerate cases.

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	}
+}
+	
+template <PartialInterpreter::VisitingPolicy policy>
```
Kill the template param

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+const llvm::BasicBlock* PartialInterpreter::visitBasicBlock(const llvm::BasicBlock* BB, const llvm::BasicBlock* from)
+{
+	assert(BB);
+	using namespace llvm;
```
This is unusual

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+template <PartialInterpreter::VisitingPolicy policy>
+const llvm::BasicBlock* PartialInterpreter::visitBasicBlock(const llvm::BasicBlock* BB, const llvm::BasicBlock* from)
+{
+	assert(BB);
```
Not especially required, we control the callers don't we?

carlopi replied:
Yes, it's more to inform that first parameter is guaranteed to be non-null (while second might be nullptr to signal unknow). I can also just say it in a comment

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	}
+	else if (const SwitchInst *SI = dyn_cast<SwitchInst>(Term))
+	{
+		if (isValueComputed(SI->getCondition()) && (getBitMask(SI->getCondition()) == BitMask::ALL) )
```
What about the default?

carlopi replied:
Yes

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		//Non handled terminator are fine, since next being nullptr means no information will be propagated
+	}
+
+	//TODO: RAII to pop
```
What is there to pop?

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+
+class ModuleData
+{
+	std::string error;
```
This is never actually used

carlopi replied:
This is passes (as reference) when creating the Interpreter. it might be used to report some recoverable errors (that we are assuming are not happening since we filter to only a limited set)

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+class FunctionData
+{
+	const llvm::Function& F;
+	public://TODO:remove public
```
Handle the comment

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+					auto	     StoredAddr = (char*) exe->ValueAddresses->toReal(exe->getOperandValue(v).PointerVal);
+        const GlobalValue* GV = exe->getGlobalValueAtAddress(StoredAddr);
+
+	if (GV)
```
Review indentation

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	}
+	const VectorOfArgs* getSomethingToVisit()
+	{
+		if (indexCallEquivalentQueue >= callEquivalentQueue.size())
```
A more usual popped stack?

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	}
+	void cleanupBB(Function& function)
+	{
+		int BBs =0;
```
Is this ever used?

carlopi replied:
Yes, by modifyFunction

alexp-sssup replied:
I meant the variable BBs in particular

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+
+		{
+			if (visitedEdges.count(p) == 0)
+				removeEdgeBetweenBlocks(p.first, p.second);
```
Maybe it would make more sense to visit the CFG and directly removed all edges which are not part of the set of visited edges.

carlopi replied:
I reworked the iteration integrating with other parts, but I think it's cleaner if the call to removeEdgeBetweenBlocks are done explicitly

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+
+		if ((numberVisitedBBs != numberBBs) || (Edges.size() != numberVisitedEdges))
+		{
+			if (emitStats)
```
Either use the standard infrastructure or avoid this altogether

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		llvm::DenseSet<const llvm::BasicBlock*> visitedBBs;
+		visitedBBs.insert(&F.getEntryBlock());
+		for (const auto& pair : visitedEdges)
+			visitedBBs.insert(pair.second);
```
Only the count is used, avoid populating the set

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		{
+			numberBBs++;
+			for (const BasicBlock* p : predecessors(&x))
+				Edges.insert({&x, p});
```
set is not really used

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		for (const BasicBlock& x : F)
+		{
+			numberBBs++;
+			for (const BasicBlock* p : predecessors(&x))
```
Visiting successors is more efficient

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+namespace llvm{
+
+
+class SubGraph;
```
Do we actually need all this complexity?

carlopi replied:
Added comment explaining the purpose of this (that we need to simplify SCCs)

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+{
+	FunctionData& data;
+	DeterministicBBSet inner;
+	bool multiHead {false};
```
Initialize everything in constructor, this mix is confusing

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	//TODO(carlo): an optmization might be having from be a set<BasicBlock>, conserving the phi that are equals
+	BasicBlockGroupData* parent;
+	std::deque<BasicBlockGroupData> subGroups;
+	std::map<const llvm::BasicBlock*, int> subGroupsID;
```
The purpose of these IDs is unclear

carlopi replied:
Those are a reverse map BB->index of children, added explanation + renamed

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+
+			return false;
+		}
+		else if (visitingAll)
```
It looks like it would be better not to call notifysuccessor in the visitingAll case. It may be possible to remove the flag completely

carlopi replied:
notifySuccessor has to be visited, because in the logic of the Tree of Subsets (https://github.com/leaningtech/cheerp-compiler/pull/64#discussion_r809416633) visitAll means "all basic blocks at the current level are considered visited" (and so you can skip the actual visit there), but notifySuccessors might still point to siblings somewhere else in the tree, and notify successor is the way to propagate the information. It could be simplifyicable by dropping visitngAll entierly, and probably it's just overall clearer.

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+			assert(parent);
+			//It should handled by the parent SCC
+
+			const bool notified = parent->notifySuccessor(from, succ, iter);
```
Unsure about this recursive calling, what is the idea?

carlopi replied:
I will add explainer. Idea is there is that this is basically a tree, with the root containing everything, and going down there are basically subsets of the BBs. If we start at a given depth, and we loop for node succ, and it's not part of the current set of BB, then has to be up in the hierarchy.
I will write down this properly (setting as resolved when I have done)

carlopi replied:
Added explainer

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		if (start)
+			assert(start->getParent() == data.getFunction());
+	}
+	BasicBlockGroupData(FunctionData& data)
```
Are these constructors actually used?

carlopi replied:
Yes, I checked

carlopi replied:
Adding a comment on the with tricky behaviour

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+			visitAll();
+			return;
+		}
+		if (!start)
```
How can this happen?

carlopi replied:
Added comment, basically say START if (something)-> A else B, and you can prove it's always A, then B will be visited but start will still nullptr. (Here START, A and B are generic SCC, not just blocks)

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		if (iterationCount >= 100)
+		{
+
+			data.visitCounter[start] = 200;
```
?

carlopi replied:
Killed

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+		if (shouldCleanUp())
+		{
+			//TODO(carlo): something should be here?
+		}
```
Well, it's a good time to figure this out isn't it?

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+					equivalentArg = false;
+			}
+
+			if (equivalentArg && candidate)
```
Can you verify if this ever triggers? I would think that the rest of the pipeline would have done this forwarding already

carlopi replied:
It triggers, but it's minor and muddies the waters, removed

carlopi commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1393 @@
+	{
+		visitedEdges.insert({from, to});
+	}
+	VectorOfArgs getArguments(const llvm::CallBase* callBase, const llvm::CallBase* parent, PartialInterpreter* exe)
```
Cleaned-up. This was infrastructure for something else dropped, cleaning up is the right choice

carlopi commented:
I pushed the current state, reworked both the infrastructure and the main file according to the review, added comments in the code, and factored out / simplified whatever was not central.

On my todo list currently there is:
 - moving to CheerpWriter folder
 - modifying the logic that bump the alignment on Global Variable to take note of which GV needs alignment to be modified + return ALIGN8 irrespective of the GV current alignment, and modify it only later (this would be more in line with first analyze, then do modifications, and I believe algorithm is clearer to reason about if the state of the module is completely frozen)

carlopi commented:
Possibly I would also squash everything, but state of the code should be more or less frozen on my side, so @alexp-sssup feel free to plan another review as you prefer.

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1440 @@
+	// Auxiliary map to determin whether a value can be queried to the Interpreter and how many bits are actually known
+	std::deque<llvm::DenseMap<const llvm::Value*, BitMask>> stronglyKnownBits;
+
+	GenericValue getOperandValueExplicit(Value* V, ExecutionContext& EC)
```
Remove this forward method, it is now useless

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1440 @@
+			immutableLoadIntervals.push_back({(long long)Ptr, ((long long)Ptr)+getDataLayout().getTypeAllocSize(GV->getInitializer()->getType())});
+		}
+
+		std::sort(immutableLoadIntervals.begin(), immutableLoadIntervals.end());
```
The ordering is currently unused

carlopi replied:
Removed sorting, added comment on meaning & lifetime of immutableLoadIntervals, can be added back as optimization later

alexp-sssup commented on llvm/lib/CheerpUtils/PartialExecuter.cpp:
```diff
@@ -0,0 +1,1440 @@
+
+				if (BI.isConditional())
+				{
+					if (isValueComputed(BI.getCondition()))
```
I am missing something, I remember bitmasks checks here that seemed to make sense before. Have they been moved elsewhere or removed? And if removed, why?

carlopi replied:
Handling of terminators is duplicated in 2 different places, findNextBasicBlock (no checks on known bits since they are done before-hand) AND PartialInterpreter::visitBasicBlock (with checks on known bits). I will unify them, probably by always doing the checks (otherwise passing a parameter no_checks, but seems overly complex)
